File: java/src/main/java/com/sketches_oxide/membership/CountingBloomFilter.java
package com.sketches_oxide.membership;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * CountingBloomFilter - Probabilistic Membership with Deletion Support.
 *
 * An extension of the classic Bloom filter that uses counters instead of bits,
 * enabling item deletion. Each position stores a count rather than a single bit,
 * allowing decrement operations when items are removed.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Uses k hash functions mapping to m counter positions</li>
 *   <li>Insert: increment all k counter positions</li>
 *   <li>Delete: decrement all k counter positions</li>
 *   <li>Query: check if all k positions have count &gt; 0</li>
 *   <li>Counters typically use 4 bits (max count 15) to limit overflow</li>
 * </ul>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Dynamic membership with additions and deletions</li>
 *   <li>Network routing tables with route removal</li>
 *   <li>Cache eviction tracking</li>
 *   <li>Session management with timeouts</li>
 * </ul>
 *
 * <p><strong>Trade-offs vs Standard Bloom:</strong>
 * <ul>
 *   <li>Space: 3-4x larger (counters vs bits)</li>
 *   <li>Supports deletion (standard Bloom does not)</li>
 *   <li>Risk of counter overflow on high-frequency items</li>
 *   <li>Same FPR guarantees when not overflowed</li>
 * </ul>
 *
 * <p><strong>Warning:</strong> Deleting items that were never inserted can cause
 * false negatives. Only delete items you are certain were previously inserted.
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (CountingBloomFilter cbf = new CountingBloomFilter(10000, 0.01)) {
 *     // Add items
 *     cbf.insert("session-123".getBytes());
 *     cbf.insert("session-456".getBytes());
 *
 *     // Check membership
 *     System.out.println(cbf.contains("session-123".getBytes())); // true
 *
 *     // Remove expired session
 *     cbf.remove("session-123".getBytes());
 *     System.out.println(cbf.contains("session-123".getBytes())); // false (probably)
 * }
 * </pre>
 *
 * @see BloomFilter
 * @see <a href="https://www.eecs.harvard.edu/~michaelm/postscripts/rsa2000.pdf">Counting Bloom Filters</a>
 */
public final class CountingBloomFilter extends NativeSketch implements MergeableSketch<CountingBloomFilter> {

    private final long n;
    private final double fpr;

    /**
     * Create a new CountingBloomFilter.
     *
     * @param n expected number of items (peak occupancy, accounting for deletions)
     * @param fpr desired false positive rate (0 &lt; fpr &lt; 1)
     * @throws IllegalArgumentException if n &lt;= 0 or fpr not in (0, 1)
     * @throws RuntimeException if native memory allocation fails
     */
    public CountingBloomFilter(long n, double fpr) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive, got: " + n);
        }
        if (fpr <= 0 || fpr >= 1) {
            throw new IllegalArgumentException("fpr must be in range (0, 1), got: " + fpr);
        }

        this.n = n;
        this.fpr = fpr;
        this.nativePtr = SketchOxideNative.countingbloomfilter_new(n, fpr);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate CountingBloomFilter");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private CountingBloomFilter(long nativePtr, long n, double fpr) {
        this.nativePtr = nativePtr;
        this.n = n;
        this.fpr = fpr;
    }

    /**
     * Insert an item into the filter.
     *
     * Increments all k counter positions for this item.
     *
     * @param item the item to insert (as raw bytes)
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.countingbloomfilter_insert(nativePtr, item);
    }

    /**
     * Insert a string item into the filter.
     *
     * @param item the string to insert
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void insert(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        insert(item.getBytes());
    }

    /**
     * Remove an item from the filter.
     *
     * Decrements all k counter positions for this item.
     *
     * <p><strong>Warning:</strong> Only remove items that were previously inserted.
     * Removing items that were never inserted can cause false negatives for
     * other items that share counter positions.
     *
     * @param item the item to remove (as raw bytes)
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void remove(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.countingbloomfilter_remove(nativePtr, item);
    }

    /**
     * Remove a string item from the filter.
     *
     * @param item the string to remove
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public void remove(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        remove(item.getBytes());
    }

    /**
     * Check if an item might be in the filter.
     *
     * Returns true if all k counter positions have count &gt; 0.
     * Returns false if any counter position is 0 (definitely not in set).
     *
     * @param item the item to check (as raw bytes)
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.countingbloomfilter_contains(nativePtr, item);
    }

    /**
     * Check if a string item might be in the filter.
     *
     * @param item the string to check
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed
     * @throws NullPointerException if item is null
     */
    public boolean contains(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return contains(item.getBytes());
    }

    /**
     * Merge another CountingBloomFilter into this one.
     *
     * After merging, this filter contains the union of both filters.
     * Counter values are added together.
     *
     * <p><strong>Note:</strong> Merging is typically not supported for counting
     * bloom filters due to counter semantics. This method may have implementation-
     * specific behavior.
     *
     * @param other the filter to merge into this one
     * @throws IllegalStateException if either filter is closed
     * @throws IllegalArgumentException if parameters don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(CountingBloomFilter other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.n != other.n || Math.abs(this.fpr - other.fpr) > 1e-9) {
            throw new IllegalArgumentException(
                    "Cannot merge CountingBloomFilters with different parameters: " +
                            String.format("(n=%d, fpr=%.6f) vs (n=%d, fpr=%.6f)",
                                    this.n, this.fpr, other.n, other.fpr));
        }

        // Note: CountingBloomFilter merge semantics may vary by implementation
        // This follows the native library's merge behavior
        throw new UnsupportedOperationException(
                "CountingBloomFilter merge is not supported - counter semantics are ambiguous");
    }

    /**
     * Serialize the filter to binary format.
     *
     * @return binary representation of the filter
     * @throws IllegalStateException if the filter has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.countingbloomfilter_serialize(nativePtr);
    }

    /**
     * Deserialize a CountingBloomFilter from binary format.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new CountingBloomFilter instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static CountingBloomFilter deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.countingbloomfilter_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized CountingBloomFilter data");
        }

        return new CountingBloomFilter(ptr, 100, 0.01);
    }

    /**
     * Get the expected number of items.
     *
     * @return the n parameter used at creation
     */
    public long getN() {
        return n;
    }

    /**
     * Get the target false positive rate.
     *
     * @return the fpr parameter used at creation
     */
    public double getFpr() {
        return fpr;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.countingbloomfilter_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("CountingBloomFilter(n=%d, fpr=%.6f)", n, fpr);
        }
        return "CountingBloomFilter(closed)";
    }
}

File: dotnet/SketchOxide/src/Membership/CountingBloomFilter.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Membership;

/// <summary>
/// Counting Bloom Filter with support for element deletion.
///
/// A variant of Bloom filter that uses counters instead of single bits,
/// allowing elements to be removed from the filter. Each position stores
/// a counter that is incremented on insert and decremented on remove.
/// </summary>
/// <remarks>
/// <para>
/// The Counting Bloom filter provides:
/// - Support for element deletion (unlike standard Bloom filter)
/// - Same probabilistic guarantees as standard Bloom filter
/// - Higher memory usage (typically 4x due to counter storage)
/// - Risk of counter overflow for extremely frequent elements
/// </para>
/// <para>
/// Ideal for applications where:
/// - Elements need to be removed from the set
/// - Dynamic membership is required
/// - Memory overhead is acceptable
/// </para>
/// <para>
/// Warning: Removing an element that was never inserted can cause false negatives.
/// Always ensure elements are only removed after being inserted.
/// </para>
/// </remarks>
public sealed class CountingBloomFilter : NativeSketch
{
    private readonly ulong _size;
    private readonly double _fpr;

    /// <summary>
    /// Creates a new Counting Bloom filter with the specified expected number of elements and false positive rate.
    /// </summary>
    /// <param name="expectedElements">Expected number of elements to be inserted. Must be greater than 0.</param>
    /// <param name="falsePositiveRate">Desired false positive rate (e.g., 0.01 for 1%). Must be in range (0, 1).</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if expectedElements is 0 or falsePositiveRate is outside (0, 1).</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public CountingBloomFilter(ulong expectedElements, double falsePositiveRate)
    {
        if (expectedElements == 0)
            throw new ArgumentOutOfRangeException(nameof(expectedElements), expectedElements, "Expected elements must be greater than 0");
        if (falsePositiveRate <= 0 || falsePositiveRate >= 1)
            throw new ArgumentOutOfRangeException(nameof(falsePositiveRate), falsePositiveRate, "False positive rate must be in range (0, 1)");

        _size = expectedElements;
        _fpr = falsePositiveRate;
        NativePtr = SketchOxideNative.countingbloomfilter_new(expectedElements, falsePositiveRate);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native CountingBloomFilter");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private CountingBloomFilter(ulong size, double fpr, nuint ptr)
    {
        _size = size;
        _fpr = fpr;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the expected number of elements this filter was configured for.
    /// </summary>
    public ulong ExpectedElements
    {
        get
        {
            CheckAlive();
            return _size;
        }
    }

    /// <summary>
    /// Gets the target false positive rate.
    /// </summary>
    public double FalsePositiveRate
    {
        get
        {
            CheckAlive();
            return _fpr;
        }
    }

    /// <summary>
    /// Inserts an element into the filter, incrementing its counters.
    /// </summary>
    /// <param name="data">The bytes representing the element to insert.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Insert(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.countingbloomfilter_insert(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Inserts a string element into the filter.
    /// </summary>
    /// <param name="value">The string to insert.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Insert(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Insert(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Tests whether an element might be in the set.
    /// </summary>
    /// <param name="data">The bytes representing the element to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.countingbloomfilter_contains(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Tests whether a string element might be in the set.
    /// </summary>
    /// <param name="value">The string to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public bool Contains(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Contains(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Removes an element from the filter by decrementing its counters.
    /// </summary>
    /// <param name="data">The bytes representing the element to remove.</param>
    /// <remarks>
    /// Warning: Removing an element that was never inserted can cause false negatives
    /// for other elements. Only remove elements that were previously inserted.
    /// </remarks>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Remove(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        unsafe
        {
            fixed (byte* ptr = data)
            {
                SketchOxideNative.countingbloomfilter_remove(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Removes a string element from the filter.
    /// </summary>
    /// <param name="value">The string to remove.</param>
    /// <remarks>
    /// Warning: Removing an element that was never inserted can cause false negatives
    /// for other elements. Only remove elements that were previously inserted.
    /// </remarks>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public void Remove(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Remove(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the filter to a byte array.
    /// </summary>
    /// <returns>Serialized filter bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.countingbloomfilter_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a Counting Bloom filter from a byte array.
    /// </summary>
    /// <param name="data">Serialized filter bytes.</param>
    /// <param name="expectedElements">The expected elements parameter used when creating the original filter.</param>
    /// <param name="falsePositiveRate">The false positive rate parameter used when creating the original filter.</param>
    /// <returns>A new CountingBloomFilter instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static CountingBloomFilter Deserialize(byte[] data, ulong expectedElements, double falsePositiveRate)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.countingbloomfilter_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize CountingBloomFilter: invalid data");

        return new CountingBloomFilter(expectedElements, falsePositiveRate, ptr);
    }

    /// <summary>
    /// Returns a string representation of the filter.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "CountingBloomFilter(disposed)";
        return $"CountingBloomFilter(expectedElements={_size}, fpr={_fpr:P2})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.countingbloomfilter_free(NativePtr);
            NativePtr = 0;
        }
    }
}
